File: internal/tools/text.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nerveband/ai-happy-design/internal/figma"
)

// RegisterTextTool registers the "text" tool for text node creation and manipulation.
func RegisterTextTool(s *server.MCPServer, commander *figma.Commander) {
	tool := mcp.NewTool("text",
		mcp.WithDescription("Text operations: create text nodes, change content, font, size, weight, color, alignment, spacing, case, decoration, and more."),
		mcp.WithString("action", mcp.Required(), mcp.Description("Action to perform"),
			mcp.Enum("create", "set_content", "set_font", "set_size", "set_weight", "set_color", "set_align", "set_spacing", "set_case", "set_decoration", "get_segments", "load_font", "set_style_id", "list_fonts")),
		mcp.WithString("nodeId", mcp.Description("Target text node ID")),
		mcp.WithString("text", mcp.Description("Text content")),
		mcp.WithString("fontFamily", mcp.Description("Font family name (e.g. Inter, Roboto)")),
		mcp.WithString("fontStyle", mcp.Description("Font style (e.g. Regular, Bold, Italic)")),
		mcp.WithNumber("fontSize", mcp.Description("Font size in pixels")),
		mcp.WithString("fontWeight", mcp.Description("Font weight (e.g. 400, 700)")),
		mcp.WithString("color", mcp.Description("Text color as hex string")),
		mcp.WithString("textAlign", mcp.Description("Text alignment: LEFT, CENTER, RIGHT, JUSTIFIED"),
			mcp.Enum("LEFT", "CENTER", "RIGHT", "JUSTIFIED")),
		mcp.WithNumber("letterSpacing", mcp.Description("Letter spacing value")),
		mcp.WithString("letterSpacingUnit", mcp.Description("Letter spacing unit: PIXELS or PERCENT"),
			mcp.Enum("PIXELS", "PERCENT")),
		mcp.WithNumber("lineHeight", mcp.Description("Line height value")),
		mcp.WithNumber("paragraphSpacing", mcp.Description("Paragraph spacing value")),
		mcp.WithString("textCase", mcp.Description("Text case: ORIGINAL, UPPER, LOWER, TITLE"),
			mcp.Enum("ORIGINAL", "UPPER", "LOWER", "TITLE")),
		mcp.WithString("textDecoration", mcp.Description("Text decoration: NONE, UNDERLINE, STRIKETHROUGH"),
			mcp.Enum("NONE", "UNDERLINE", "STRIKETHROUGH")),
		mcp.WithString("styleId", mcp.Description("Text style ID to apply")),
		mcp.WithNumber("x", mcp.Description("X position for new text")),
		mcp.WithNumber("y", mcp.Description("Y position for new text")),
		mcp.WithString("parentId", mcp.Description("Parent node ID")),
		mcp.WithString("name", mcp.Description("Name for the text node")),
		mcp.WithNumber("width", mcp.Description("Text box width for wrapping")),
		mcp.WithNumber("opacity", mcp.Description("Opacity from 0 to 1")),
		mcp.WithString("lineHeightUnit", mcp.Description("Line height unit: PIXELS, PERCENT, or AUTO"),
			mcp.Enum("PIXELS", "PERCENT", "AUTO")),
		mcp.WithString("layoutSizingHorizontal", mcp.Description("Horizontal sizing in auto-layout: FIXED, HUG, or FILL"),
			mcp.Enum("FIXED", "HUG", "FILL")),
		mcp.WithString("layoutSizingVertical", mcp.Description("Vertical sizing in auto-layout: FIXED, HUG, or FILL"),
			mcp.Enum("FIXED", "HUG", "FILL")),
		mcp.WithNumber("layoutGrow", mcp.Description("Layout grow: 0 or 1")),
		mcp.WithString("layoutAlign", mcp.Description("Layout alignment: STRETCH or INHERIT"),
			mcp.Enum("STRETCH", "INHERIT")),
	)

	s.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := req.GetArguments()
		action := getStringArg(args, "action", "")

		switch action {
		case "create":
			params := map[string]interface{}{
				"text":       getStringArg(args, "text", "Hello"),
				"x":          getFloat64Arg(args, "x", 0),
				"y":          getFloat64Arg(args, "y", 0),
				"fontFamily": getStringArg(args, "fontFamily", "Inter"),
				"fontStyle":  getStringArg(args, "fontStyle", "Regular"),
				"fontSize":   getFloat64Arg(args, "fontSize", 16),
				"parentId":   getStringArg(args, "parentId", ""),
				"name":       getStringArg(args, "name", ""),
			}
			// Forward optional params only if explicitly provided
			if hasArg(args, "color") {
				params["color"] = getStringArg(args, "color", "")
			}
			if hasArg(args, "width") {
				params["width"] = getFloat64Arg(args, "width", 0)
			}
			if hasArg(args, "textAlign") {
				params["textAlignHorizontal"] = getStringArg(args, "textAlign", "")
			}
			if hasArg(args, "lineHeight") {
				params["lineHeight"] = getFloat64Arg(args, "lineHeight", 0)
			}
			if hasArg(args, "lineHeightUnit") {
				params["lineHeightUnit"] = getStringArg(args, "lineHeightUnit", "")
			}
			if hasArg(args, "letterSpacing") {
				params["letterSpacing"] = getFloat64Arg(args, "letterSpacing", 0)
			}
			if hasArg(args, "letterSpacingUnit") {
				params["letterSpacingUnit"] = getStringArg(args, "letterSpacingUnit", "")
			}
			if hasArg(args, "textCase") {
				params["textCase"] = getStringArg(args, "textCase", "")
			}
			if hasArg(args, "opacity") {
				params["opacity"] = getFloat64Arg(args, "opacity", 1)
			}
			if hasArg(args, "layoutSizingHorizontal") {
				params["layoutSizingHorizontal"] = getStringArg(args, "layoutSizingHorizontal", "")
			}
			if hasArg(args, "layoutSizingVertical") {
				params["layoutSizingVertical"] = getStringArg(args, "layoutSizingVertical", "")
			}
			if hasArg(args, "layoutGrow") {
				params["layoutGrow"] = getFloat64Arg(args, "layoutGrow", 0)
			}
			if hasArg(args, "layoutAlign") {
				params["layoutAlign"] = getStringArg(args, "layoutAlign", "")
			}
			return sendCommand(commander, "create_text", params)

		case "set_content":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_text_content", map[string]interface{}{
				"nodeId": nodeId,
				"text":   getStringArg(args, "text", ""),
			})

		case "set_font":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_font_name", map[string]interface{}{
				"nodeId":     nodeId,
				"fontFamily": getStringArg(args, "fontFamily", "Inter"),
				"fontStyle":  getStringArg(args, "fontStyle", "Regular"),
			})

		case "set_size":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_font_size", map[string]interface{}{
				"nodeId":   nodeId,
				"fontSize": getFloat64Arg(args, "fontSize", 16),
			})

		case "set_weight":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_font_weight", map[string]interface{}{
				"nodeId":     nodeId,
				"fontWeight": getStringArg(args, "fontWeight", "400"),
			})

		case "set_color":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_fill_color", map[string]interface{}{
				"nodeId": nodeId,
				"color":  getStringArg(args, "color", "#000000"),
			})

		case "set_align":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_text_align", map[string]interface{}{
				"nodeId":    nodeId,
				"textAlign": getStringArg(args, "textAlign", "LEFT"),
			})

		case "set_spacing":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_text_spacing", map[string]interface{}{
				"nodeId":           nodeId,
				"letterSpacing":    getFloat64Arg(args, "letterSpacing", 0),
				"lineHeight":       getFloat64Arg(args, "lineHeight", 0),
				"paragraphSpacing": getFloat64Arg(args, "paragraphSpacing", 0),
			})

		case "set_case":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_text_case", map[string]interface{}{
				"nodeId":   nodeId,
				"textCase": getStringArg(args, "textCase", "ORIGINAL"),
			})

		case "set_decoration":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_text_decoration", map[string]interface{}{
				"nodeId":         nodeId,
				"textDecoration": getStringArg(args, "textDecoration", "NONE"),
			})

		case "get_segments":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "get_styled_text_segments", map[string]interface{}{
				"nodeId": nodeId,
			})

		case "load_font":
			return sendCommand(commander, "load_font_async", map[string]interface{}{
				"fontFamily": getStringArg(args, "fontFamily", "Inter"),
				"fontStyle":  getStringArg(args, "fontStyle", "Regular"),
			})

		case "set_style_id":
			nodeId, errResult := requireStringArg(args, "nodeId")
			if errResult != nil {
				return errResult, nil
			}
			return sendCommand(commander, "set_text_style_id", map[string]interface{}{
				"nodeId":  nodeId,
				"styleId": getStringArg(args, "styleId", ""),
			})

		case "list_fonts":
			params := map[string]interface{}{}
			if hasArg(args, "fontFamily") {
				params["family"] = getStringArg(args, "fontFamily", "")
			}
			return sendCommand(commander, "list_fonts", params)

		default:
			return mcp.NewToolResultError(fmt.Sprintf("unknown text action: %s", action)), nil
		}
	})
}

File: plugin/src/handlers/text.ts
import { loadFont, loadNodeFonts, resolveFontFamily } from '../utils/fonts';
import { getTextNodeById, getParentById } from '../utils/getNode';
import { resolveStableId } from '../utils/stableId';

function parseHexColor(color: any, fallback = { r: 0, g: 0, b: 0, a: 1 }) {
  if (color && typeof color === 'object' && typeof color.r === 'number') {
    return {
      r: color.r,
      g: color.g,
      b: color.b,
      a: typeof color.a === 'number' ? color.a : 1,
    };
  }

  if (typeof color !== 'string') return fallback;
  const raw = color.trim().replace(/^#/, '');
  if (!(raw.length === 3 || raw.length === 6 || raw.length === 8)) return fallback;

  const hex = raw.length === 3
    ? raw.split('').map(ch => ch + ch).join('')
    : raw;
  const hasAlpha = hex.length === 8;
  const n = parseInt(hex, 16);
  if (Number.isNaN(n)) return fallback;

  return {
    r: ((n >> (hasAlpha ? 24 : 16)) & 0xff) / 255,
    g: ((n >> (hasAlpha ? 16 : 8)) & 0xff) / 255,
    b: ((n >> (hasAlpha ? 8 : 0)) & 0xff) / 255,
    a: hasAlpha ? (n & 0xff) / 255 : 1,
  };
}

export async function handleText(action: string, params: any): Promise<any> {
  switch (action) {
    case 'create_text':
    case 'add_text':
    case 'add':
    case 'new':
    case 'create': return createText(params);
    case 'set_text':
    case 'update_text':
    case 'edit':
    case 'set_content': return setContent(params);
    case 'font':
    case 'set_font_family':
    case 'set_font': return setFont(params);
    case 'size':
    case 'set_font_size':
    case 'font_size':
    case 'set_size': return setSize(params);
    case 'weight':
    case 'set_font_weight':
    case 'font_weight':
    case 'bold':
    case 'set_weight': return setWeight(params);
    case 'color':
    case 'set_text_color':
    case 'text_color':
    case 'set_color': return setColor(params);
    case 'align':
    case 'set_text_align':
    case 'alignment':
    case 'set_align': return setAlign(params);
    case 'set_spacing': return setSpacing(params);
    case 'set_line_height': return setLineHeight(params);
    case 'set_letter_spacing': return setLetterSpacing(params);
    case 'set_decoration': return setDecoration(params);
    case 'set_case': return setCase(params);
    case 'set_paragraph_spacing': return setParagraphSpacing(params);
    case 'content':
    case 'get_text':
    case 'read':
    case 'get_content': return getContent(params);
    case 'get_segments': return getSegments(params);
    case 'load_font': return loadFontAction(params);
    case 'set_style_id': return setStyleId(params);
    case 'list_fonts':
    case 'available_fonts':
    case 'list_available_fonts': return listFonts(params);
    case 'set_range_style':
    case 'range_style':
    case 'style_ranges': return setRangeStyle(params);
    default: throw new Error('Unknown text action: ' + action + '. Available: create, set_content, set_font, set_size, set_weight, set_color, set_align, set_spacing, set_line_height, set_letter_spacing, set_decoration, set_case, set_paragraph_spacing, get_content, get_segments, load_font, set_style_id, list_fonts, set_range_style');
  }
}

async function createText(params: any) {
  const content = params.content ?? params.text ?? '';
  const fontFamily = params.fontFamily ?? params.family ?? 'Inter';
  const fontStyle = params.fontStyle ?? params.style ?? 'Regular';
  const fontSize = params.fontSize ?? params.size ?? 16;
  const width = params.width;

  const family = resolveFontFamily(fontFamily);
  await loadFont(family, fontStyle);

  const text = figma.createText();
  text.fontName = { family, style: fontStyle };
  text.fontSize = fontSize;
  text.x = params.x ?? 0;
  text.y = params.y ?? 0;
  if (params.name) text.name = params.name;
  text.characters = String(content);

  if (params.color) {
    const c = parseHexColor(params.color);
    text.fills = [{
      type: 'SOLID',
      color: { r: c.r, g: c.g, b: c.b },
      opacity: c.a,
    }];
  }

  if (width !== undefined) {
    text.resize(width, text.height);
    text.textAutoResize = 'HEIGHT';
  } else {
    text.textAutoResize = 'WIDTH_AND_HEIGHT';
  }

  var container: BaseNode & ChildrenMixin;
  if (params.parentId) {
    const parent = await getParentById(params.parentId);
    if (parent) {
      parent.appendChild(text);
      container = parent;
    } else {
      container = figma.currentPage;
    }
  } else {
    container = figma.currentPage;
  }

  var stableId = await resolveStableId(text, container);

  // Auto-layout child properties
  var parentHasAutoLayout = container && 'layoutMode' in container && (container as any).layoutMode !== 'NONE';

  // Auto-fix: text in auto-layout with HEIGHT resize needs layoutSizingVertical = HUG
  if (parentHasAutoLayout && text.textAutoResize === 'HEIGHT' && params.layoutSizingVertical === undefined) {
    (text as any).layoutSizingVertical = 'HUG';
  }

  if (params.layoutSizingHorizontal !== undefined) (text as any).layoutSizingHorizontal = params.layoutSizingHorizontal;
  if (params.layoutSizingVertical !== undefined) (text as any).layoutSizingVertical = params.layoutSizingVertical;
  if (params.layoutAlign !== undefined) (text as any).layoutAlign = params.layoutAlign;
  if (params.layoutGrow !== undefined) (text as any).layoutGrow = params.layoutGrow;

  // Forward textAlignHorizontal on create
  var textAlign = params.textAlignHorizontal ?? params.textAlign;
  if (textAlign) text.textAlignHorizontal = textAlign;

  // Forward lineHeight on create
  if (params.lineHeight !== undefined) {
    var lhUnit = params.lineHeightUnit ?? 'PERCENT';
    text.lineHeight = lhUnit === 'AUTO'
      ? { unit: 'AUTO' }
      : { value: params.lineHeight, unit: lhUnit === 'PERCENT' ? 'PERCENT' : 'PIXELS' };
  }

  if (params.letterSpacing !== undefined) {
    var lsUnit = params.letterSpacingUnit ?? 'PIXELS';
    text.letterSpacing = { value: params.letterSpacing, unit: lsUnit === 'PERCENT' ? 'PERCENT' : 'PIXELS' };
  }

  if (params.textCase !== undefined) {
    text.textCase = params.textCase;
  }

  if (params.opacity !== undefined) {
    text.opacity = Math.max(0, Math.min(1, params.opacity));
  }

  return { id: stableId, name: text.name, type: text.type, width: text.width, height: text.height, parentId: params.parentId || figma.currentPage.id };
}

async function setContent(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);
  node.characters = String(params.content ?? params.text ?? '');
  return { id: node.id, name: node.name, characters: node.characters };
}

async function setFont(params: any) {
  const node = await getTextNodeById(params.nodeId);
  const fontFamily = params.fontFamily ?? params.family ?? 'Inter';
  const fontStyle = params.fontStyle ?? params.style ?? 'Regular';
  const family = resolveFontFamily(fontFamily);
  await loadFont(family, fontStyle);

  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeFontName(params.rangeStart, params.rangeEnd, { family, style: fontStyle });
  } else {
    await loadNodeFonts(node);
    node.fontName = { family, style: fontStyle };
  }
  return { id: node.id, name: node.name };
}

async function setSize(params: any) {
  const node = await getTextNodeById(params.nodeId);
  const size = params.size ?? params.fontSize;
  await loadNodeFonts(node);

  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeFontSize(params.rangeStart, params.rangeEnd, size);
  } else {
    node.fontSize = size;
  }
  return { id: node.id, name: node.name };
}

async function setWeight(params: any) {
  const node = await getTextNodeById(params.nodeId);
  const weight = params.weight ?? params.fontWeight;

  const weightMap: Record<number, string> = {
    100: 'Thin',
    200: 'Extra Light',
    300: 'Light',
    400: 'Regular',
    500: 'Medium',
    600: 'Semi Bold',
    700: 'Bold',
    800: 'Extra Bold',
    900: 'Black',
  };
  const numericWeight = typeof weight === 'string' ? parseInt(weight, 10) : weight;
  const styleName = typeof weight === 'string' && Number.isNaN(numericWeight)
    ? weight
    : (weightMap[numericWeight] || 'Regular');

  const currentFont = node.fontName;
  const family = currentFont === figma.mixed ? 'Inter' : currentFont.family;
  await loadFont(family, styleName);

  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeFontName(params.rangeStart, params.rangeEnd, { family, style: styleName });
  } else {
    node.fontName = { family, style: styleName };
  }
  return { id: node.id, name: node.name };
}

async function setColor(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const c = parseHexColor(params.color);
  const fill: SolidPaint = {
    type: 'SOLID',
    color: { r: c.r, g: c.g, b: c.b },
    opacity: c.a,
  };

  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeFills(params.rangeStart, params.rangeEnd, [fill]);
  } else {
    node.fills = [fill];
  }
  return { id: node.id, name: node.name };
}

async function setAlign(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const horizontal = params.horizontal ?? params.textAlign ?? params.textAlignHorizontal;
  const vertical = params.vertical ?? params.textAlignVertical;
  if (horizontal) node.textAlignHorizontal = horizontal;
  if (vertical) node.textAlignVertical = vertical;
  return { id: node.id, name: node.name };
}

async function setSpacing(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  if (params.letterSpacing !== undefined) {
    const unit = params.letterSpacingUnit ?? 'PIXELS';
    node.letterSpacing = { value: params.letterSpacing, unit: unit === 'PERCENT' ? 'PERCENT' : 'PIXELS' };
  }
  if (params.lineHeight !== undefined) {
    const unit = params.lineHeightUnit ?? 'PIXELS';
    node.lineHeight = unit === 'AUTO'
      ? { unit: 'AUTO' }
      : { value: params.lineHeight, unit: unit === 'PERCENT' ? 'PERCENT' : 'PIXELS' };
  }
  if (params.paragraphSpacing !== undefined) {
    node.paragraphSpacing = params.paragraphSpacing;
  }

  return { id: node.id, name: node.name };
}

async function setLineHeight(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const value = params.value ?? params.lineHeight;
  const unit = params.unit ?? params.lineHeightUnit ?? 'PIXELS';
  const lineHeight: LineHeight = unit === 'AUTO'
    ? { unit: 'AUTO' }
    : { value, unit: unit === 'PERCENT' ? 'PERCENT' : 'PIXELS' };

  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeLineHeight(params.rangeStart, params.rangeEnd, lineHeight);
  } else {
    node.lineHeight = lineHeight;
  }
  return { id: node.id, name: node.name };
}

async function setLetterSpacing(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const value = params.value ?? params.letterSpacing;
  const unit = params.unit ?? params.letterSpacingUnit ?? 'PIXELS';
  const letterSpacing: LetterSpacing = {
    value,
    unit: unit === 'PERCENT' ? 'PERCENT' : 'PIXELS',
  };

  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeLetterSpacing(params.rangeStart, params.rangeEnd, letterSpacing);
  } else {
    node.letterSpacing = letterSpacing;
  }
  return { id: node.id, name: node.name };
}

async function setDecoration(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const decoration = params.decoration ?? params.textDecoration;
  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeTextDecoration(params.rangeStart, params.rangeEnd, decoration);
  } else {
    node.textDecoration = decoration;
  }
  return { id: node.id, name: node.name };
}

async function setCase(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const textCase = params.textCase ?? params.case;
  if (params.rangeStart !== undefined && params.rangeEnd !== undefined) {
    node.setRangeTextCase(params.rangeStart, params.rangeEnd, textCase);
  } else {
    node.textCase = textCase;
  }
  return { id: node.id, name: node.name };
}

async function setParagraphSpacing(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);
  node.paragraphSpacing = params.spacing ?? params.paragraphSpacing ?? 0;
  return { id: node.id, name: node.name };
}

async function getContent(params: any) {
  const node = await getTextNodeById(params.nodeId);
  return {
    id: node.id,
    name: node.name,
    characters: node.characters,
    fontSize: node.fontSize,
    fontName: node.fontName,
    textAlignHorizontal: node.textAlignHorizontal,
    textAlignVertical: node.textAlignVertical,
    lineHeight: node.lineHeight,
    letterSpacing: node.letterSpacing,
    textDecoration: node.textDecoration,
    textCase: node.textCase,
  };
}

async function getSegments(params: any) {
  const node = await getTextNodeById(params.nodeId);
  const property = params.property;
  const properties = property
    ? [property]
    : ['fontName', 'fontSize', 'fills', 'textDecoration', 'textCase', 'lineHeight', 'letterSpacing'];
  const segments = node.getStyledTextSegments(properties as any);
  return { id: node.id, name: node.name, segments };
}

async function loadFontAction(params: any) {
  const fontFamily = params.fontFamily ?? params.family ?? 'Inter';
  const fontStyle = params.fontStyle ?? params.style ?? 'Regular';
  const family = resolveFontFamily(fontFamily);
  await loadFont(family, fontStyle);
  return { loaded: true, family, style: fontStyle };
}

async function setStyleId(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await node.setTextStyleIdAsync(params.styleId ?? '');
  return { id: node.id, name: node.name, textStyleId: node.textStyleId };
}

async function listFonts(params: any) {
  const fonts = await figma.listAvailableFontsAsync();
  const familyFilter = params.family ?? params.fontFamily;

  // Group by family
  const familyMap = new Map<string, string[]>();
  for (const font of fonts) {
    if (familyFilter && !font.fontName.family.toLowerCase().includes(familyFilter.toLowerCase())) {
      continue;
    }
    const existing = familyMap.get(font.fontName.family);
    if (existing) {
      existing.push(font.fontName.style);
    } else {
      familyMap.set(font.fontName.family, [font.fontName.style]);
    }
  }

  const result: { family: string; styles: string[] }[] = [];
  for (const [family, styles] of familyMap) {
    result.push({ family, styles });
  }

  return { fonts: result, count: result.length };
}

/**
 * Resolve a range specification to {start, end} indices.
 * Exported for testing.
 */
export function resolveRange(
  range: { match?: string; start?: number; end?: number },
  text: string,
): { start: number; end: number } | null {
  let start: number;
  let end: number;

  if (typeof range.match === 'string') {
    const idx = text.indexOf(range.match);
    if (idx === -1) return null;
    start = idx;
    end = idx + range.match.length;
  } else if (typeof range.start === 'number' && typeof range.end === 'number') {
    start = range.start;
    end = range.end;
  } else {
    return null;
  }

  // Clamp to text bounds
  const len = text.length;
  start = Math.max(0, Math.min(start, len));
  end = Math.max(start, Math.min(end, len));
  if (start === end) return null;

  return { start, end };
}

async function setRangeStyle(params: any) {
  const node = await getTextNodeById(params.nodeId);
  await loadNodeFonts(node);

  const ranges: any[] = params.ranges;
  if (!Array.isArray(ranges) || ranges.length === 0) {
    throw new Error('set_range_style requires a non-empty "ranges" array');
  }

  const text = node.characters;
  let rangesApplied = 0;

  for (const range of ranges) {
    const resolved = resolveRange(range, text);
    if (!resolved) continue;

    const { start, end } = resolved;

    // --- Font name (bold / italic / fontFamily / fontStyle) ---
    if (range.bold !== undefined || range.italic !== undefined || range.fontFamily || range.fontStyle) {
      // Get current font at range start to use as base
      let baseFontName: FontName;
      const rangeFont = node.getRangeFontName(start, start + 1);
      if (rangeFont === figma.mixed) {
        const overall = node.fontName;
        baseFontName = overall === figma.mixed
          ? { family: 'Inter', style: 'Regular' }
          : overall;
      } else {
        baseFontName = rangeFont;
      }

      let family = range.fontFamily
        ? resolveFontFamily(range.fontFamily)
        : baseFontName.family;

      let style: string;
      if (range.fontStyle) {
        // Explicit fontStyle takes priority
        style = range.fontStyle;
      } else {
        // Build style from bold/italic flags
        const isBold = range.bold !== undefined ? range.bold : /Bold/i.test(baseFontName.style);
        const isItalic = range.italic !== undefined ? range.italic : /Italic/i.test(baseFontName.style);

        if (isBold && isItalic) style = 'Bold Italic';
        else if (isBold) style = 'Bold';
        else if (isItalic) style = 'Italic';
        else style = 'Regular';
      }

      await loadFont(family, style);
      node.setRangeFontName(start, end, { family, style });
    }

    // --- Font size ---
    if (range.fontSize !== undefined) {
      node.setRangeFontSize(start, end, range.fontSize);
    }

    // --- Color ---
    if (range.color !== undefined) {
      const c = parseHexColor(range.color);
      const fill: SolidPaint = {
        type: 'SOLID',
        color: { r: c.r, g: c.g, b: c.b },
        opacity: c.a,
      };
      node.setRangeFills(start, end, [fill]);
    }

    // --- Letter spacing ---
    if (range.letterSpacing !== undefined) {
      const unit = range.letterSpacingUnit ?? 'PIXELS';
      node.setRangeLetterSpacing(start, end, {
        value: range.letterSpacing,
        unit: unit === 'PERCENT' ? 'PERCENT' : 'PIXELS',
      });
    }

    // --- Line height ---
    if (range.lineHeight !== undefined) {
      const unit = range.lineHeightUnit ?? 'PIXELS';
      const lh: LineHeight = unit === 'AUTO'
        ? { unit: 'AUTO' }
        : { value: range.lineHeight, unit: unit === 'PERCENT' ? 'PERCENT' : 'PIXELS' };
      node.setRangeLineHeight(start, end, lh);
    }

    // --- Text decoration ---
    if (range.textDecoration !== undefined) {
      node.setRangeTextDecoration(start, end, range.textDecoration);
    }

    // --- Text case ---
    if (range.textCase !== undefined) {
      node.setRangeTextCase(start, end, range.textCase);
    }

    rangesApplied++;
  }

  return { id: node.id, name: node.name, rangesApplied };
}
